README
#
# $Id: README,v 1.25 2012/04/04 20:29:44 evertonm Exp $
#

INTRODUCTION

	fetchconfig is a Perl script for retrieving configuration of
	multiple devices. It has been tested under Linux and Windows,
	and currently supports a variety of devices, including Cisco
	IOS, Cisco Catalyst, Cisco ASA, Cisco PIX, FortiGate
	Firewalls, ProCurve, Alcatel Ethernet Routers (aka
	Riverstone), Dell PowerConnect Switches, Terayon 3200/3500
	CMTS, Datacom DmSwitch Switches, HP MSR Routers, Mikrotik
	Routers, Tellabs MSR Routers, Juniper EX Switch JunOS.

LICENSE

        fetchconfig - Retrieving configuration for multiple devices
        Copyright (C) 2006 Everton da Silva Marques

        fetchconfig is free software; you can redistribute it and/or
        modify it under the terms of the GNU General Public License as
        published by the Free Software Foundation; either version 2,
        or (at your option) any later version.

        fetchconfig is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
        GNU General Public License for more details.

	You should have received a copy of the GNU General Public
	License along with fetchconfig; see the file COPYING. If not,
	write to the Free Software Foundation, Inc., 51 Franklin St,
	Fifth Floor, Boston, MA 02110-1301 USA.

INSTALLATION (UNIX/Linux)

	1. fetchconfig.pl expects a Perl interpreter at /usr/bin/perl.

	2. fetchconfig.pl requires the Net::Telnet module.
	(Use "cpan install Net::Telnet" if you don't have it.)

	3. Unpack the fetchconfig tarball at a suitable
        location. Example:

	[ fetchconfig tarball available at /tmp/fetchconfig-0.0.tar.gz ]

	cd /usr/local
	tar xzf /tmp/fetchconfig-0.0.tar.gz
	ln -s fetchconfig-0.0 fetchconfig

	[ fetchconfig is now available under /usr/local/fetchconfig ]

INSTALLATION (Windows)

	1. fetchconfig.pl expects to find a Perl interpreter in the PATH.

	2. fetchconfig.pl requires the Net::Telnet module.
	(Use "cpan install Net::Telnet" if you don't have it.)

	3. Unpack the fetchconfig tarball at a suitable
	location.  From the command line, gzip and tar (links
	below) can be used.
	http://www.gzip.org
	http://gnuwin32.sourceforge.net/packages/tar.htm

	Alternatively, some commercial programs (such as WinZip)
	offer a GUI.

	Command Line Example:

	[ fetchconfig tarball available at c:\temp\fetchconfig-0.0.tar.gz ]

	c:
	cd \
	gzip -cd c:\temp\fetchconfig-0.0.tar.gz | tar xvf -
	move fetchconfig-0.0 fetchconfig

	[ fetchconfig is now available under c:\fetchconfig ]

USAGE

	1. Edit the file "device_table.example" according your needs,
        and save it as "device_table". Example:

	[ fetchconfig available under /usr/local/fetchconfig ]

	cd /usr/local/fetchconfig
	cp device_table.example device_table
	vi device_table

	2. Run fetchconfig.pl, as in the following example:

	[ fetchconfig available under /usr/local/fetchconfig ]

	cd /usr/local/fetchconfig
	./fetchconfig.pl -devices=device_table

	The fetchconfig.pl script will scan the "device_table" file,
	trying to retrieve the configuration for every device, storing
	configuration files under the "repository".

DEVICE TABLE

	The device table assigned with '-device=' switch is a text
        file which supports two types of lines:

	1. Default options lines assign default options to a device
        model. The following example adds some default options to the
        "cisco-ios" model:

	#default	model		model-default-options
	#
	default:        cisco-ios       user=backup,pass=fran,enable=jose

	2. Device lines define attributes of every device and
        optionally assign device-specific options. The following
        example define a device named "vpn-gw":

	#model		dev-unique-id	device-host	[device-options]
	#
	cisco-ios       vpn-gw          192.168.0.1     keep=10,changes_only=1

	The dev-unique-id must be:
	a) an unique identifier across all devices;
	b) a valid filesystem's directory name.

DEVICE SUPPORT

	The following devices are known to be supported.  This
	is not a comprehensive list.  Other devices and versions
	are likely to work but have not been verified.

	Model Name	Devices/OS Supported
	----------	------------------------------------------------

	cisco-cat	CatOS switches

	cisco-ios	IOS routers
			2500 series, 2600 series, 2800 series,
			CS-516 (with show_cmd=wrterm option)

	cisco-asa	Cisco ASA firewalls

	cisco-pix	Cisco PIX firewalls

	fortigate	FortiOS 3.0

	parks		Parks 209R

	riverstone	Alcatel Ethernet Router 3100 (Riverstone RS3100)

	terayon-os	Terayon 3200/3500 CMTS

	dmswitch	Datacom DmSwitch

	3com-msr	3Com MSR routers

	mikrotik	MikroTik routers

	tellabs-msr	Tellabs MSR routers

	junos		Juniper EX Switch JunOS


OPTIONS

	The modules CiscoIOS, CiscoCAT, CiscoASA and CiscoPIX
	recognize the following options:

	pass		A mandatory option, specifies the login password.

	user		Specifies the login username. It can be omitted
                        for devices not requiring a login username.

	enable		Mandatory, specifies the enable password.

	timeout		Mandatory, how long to wait for TELNET responses
                        from devices.

	fetch_timeout	Optional. If given, will override the timeout
                        value used to fetch all the configuration lines.
                        It is useful for devices which spend several
                        seconds building the configuration.

	repository	Mandatory, specifies the base directory for
                        saving the configuration files.

	keep		Mandatory, the maximum number of config files
                        to retain for the device. When this limit is
                        reached, the older files are discarded.

	changes_only	Optional. If specified as changes_only=1,
                        only new configurations are saved. Otherwise,
                        configurations are saved whenever the script
                        runs; if not defined as changes_only=1, the
			script might possibly retain multiple identical
			configuration files.

	on_fetch_run	Optional. Path to an external program to be
			called whenever the configuration is fetched.
			The following environmental variables are
			passed to the program:
			FETCHCONFIG_DEV_ID = device unique id
			FETCHCONFIG_DEV_HOST = device hostname (or IP addr)
			FETCHCONFIG_PREV = pathname to previous config,
                                           undefined if none exists
			FETCHCONFIG_CURR = pathname to current config
			Example:
			on_fetch_run=/usr/local/bin/mail_diff_to_admin.sh
			Please notice that, if changes_only=1, the
			backup file referenced to FETCHCONFIG_CURR
			will be erased right after the external program
			is finished. This option is available for all
			modules.

	timezone	Optional. If defined as timezone=hide, it will
			supress timezone information from config filename.

	filename_append_suffix
			Optional. If defined, it is appended to the
			generated filename. This option is available
			for all modules.
			Example: filename_append_suffix=.txt

	Options specific to CiscoIOS:

	show_cmd	Optional. If specified, its value will replace
			the usual "show run" command. This is necessary
                        for some older devices. It may be useful to
			send alternative commands to capture the
			configuration. To keep compatibility with
			previous versions of fetchconfig, if specified
			as show_cmd=wrterm, "write term" will be used
			in place of the usual "show run" command.

	The fortigate module recognizes the following options:
	(See above cisco-ios/cisco-cat section for descriptions.)

	user		Mandatory.
	pass		Mandatory.
	repository	Mandatory.
	keep		Mandatory.
	timeout		Mandatory.
	fetch_timeout	Optional.
	changes_only	Optional.
	on_fetch_run	Optional.
	timezone	Optional.
	filename_append_suffix Optional.

	The parks module recognizes the following options:
	(See above cisco-ios/cisco-cat section for descriptions.)

	user		Mandatory.
	pass		Mandatory.
	repository	Mandatory.
	keep		Mandatory.
	timeout		Mandatory.
	fetch_timeout	Optional.
	changes_only	Optional.
	on_fetch_run	Optional.
	timezone	Optional.
	filename_append_suffix Optional.

	Options specific to Parks:

	banner_timeout	Optional. If given, will override the timeout
                        value used to wait for the 'login:' prompt.
                        It is useful since the Parks 209R spends
 			many seconds before issuing the initial
			login prompt for TELNET connections.

	The riverstone module recognizes the following options:
	(See above cisco-ios/cisco-cat section for descriptions.)

	user		Mandatory.
	pass		Mandatory.
	enable		Mandatory.
	repository	Mandatory.
	keep		Mandatory.
	timeout		Mandatory.
	fetch_timeout	Optional.
	changes_only	Optional.
	on_fetch_run	Optional.
	timezone	Optional.
	filename_append_suffix Optional.

	The terayon-os module recognizes the following options:
	(See above cisco-ios/cisco-cat section for descriptions.)

	user		Mandatory.
	pass		Mandatory.
	enable		Mandatory.
	repository	Mandatory.
	keep		Mandatory.
	timeout		Mandatory.
	fetch_timeout	Optional.
	changes_only	Optional.
	on_fetch_run	Optional.
	timezone	Optional.
	filename_append_suffix Optional.

	The dmswitch module recognizes the following options:
	(See above cisco-ios/cisco-cat section for descriptions.)

	user		Mandatory.
	pass		Mandatory.
	repository	Mandatory.
	keep		Mandatory.
	timeout		Mandatory.
	fetch_timeout	Optional.
	changes_only	Optional.
	on_fetch_run	Optional.
	timezone	Optional.
	filename_append_suffix Optional.

	The 3com-msr module recognizes the following options:
	(See above cisco-ios/cisco-cat section for descriptions.)

	user		Mandatory.
	pass		Mandatory.
	enable		Mandatory.
	timeout		Mandatory.
	repository	Mandatory.
	keep		Mandatory.
	fetch_timeout	Optional.
	changes_only	Optional.
	on_fetch_run	Optional.
	timezone	Optional.
	filename_append_suffix Optional.

	The mikrotik module recognizes the following options:
	(See above cisco-ios/cisco-cat section for descriptions.)

	user		Mandatory.
	pass		Mandatory.
	enable		Mandatory.
	timeout		Mandatory.
	repository	Mandatory.
	keep		Mandatory.
	fetch_timeout	Optional.
	changes_only	Optional.
	on_fetch_run	Optional.
	timezone	Optional.
	filename_append_suffix Optional.

	The tellabs module recognizes the following options:
	(See above cisco-ios/cisco-cat section for descriptions.)

	user		Mandatory.
	pass		Mandatory.
	enable		Mandatory.
	timeout		Mandatory.
	repository	Mandatory.
	keep		Mandatory.
	fetch_timeout	Optional.
	changes_only	Optional.
	on_fetch_run	Optional.
	timezone	Optional.
	filename_append_suffix Optional.

	The junos module recognizes the following options:
	(See above cisco-ios/cisco-cat section for descriptions.)

	user		Mandatory.
	pass		Mandatory.
	timeout		Mandatory.
	repository	Mandatory.
	keep		Mandatory.
	fetch_timeout	Optional.
	changes_only	Optional.
	on_fetch_run	Optional.
	timezone	Optional.
	filename_append_suffix Optional.

				 -- END --



CHANGES
#
# $Id: CHANGES,v 1.51 2012/04/04 20:29:43 evertonm Exp $
#

0.22
====
+ JunOS.pm: support for Juniper JunOS.

0.21
====
+ TellabsMSR.pm: support for Tellabs MSR Routers.

0.20
====
+ CiscoPIX.pm: support for Cisco PIX.

0.19
====
+ suport for Mikrotik RB800.

0.18
====
+ support for 3Com MSR devices.

0.17
====
+ bug #25475: Stuart Schneis added compatibility for Dell 6224P, 5324, 5424

0.16
====
+ rip@devco.net added CiscoASA.pm sr #106606: Cisco ASA Support 

0.15
====
+ support equal '=' character in values for options.

0.14
====
+ option timezone=hide omits timezone name from file name.
+ added DmSwitch.pm for Datacom DmSwitch devices.

0.13
====
+ resort to strftime -%Z for tz name in systems which
  lack support for strftime %z for tz offset (Solaris 7).
+ added FETCHCONFIG_DEV_ID and FETCHCONFIG_DEV_HOST
  environmental variables to external on_fetch_run program.

0.12
====
+ option on_fetch_run to call an external program whenever
  a config backup is fetched.
+ append timezone offset to backup filename.
+ Sergey Alexanov added Terayon.pm for Old Terayon 3200/3500 CMTS.

0.11
====
+ add retrieval timing info into log.

0.10
====
+ bug #21501: Problem if device-specific-options are not specified

0.9
===
+ Slightly better device table parser in order to support
  options with blank spaces. Example: show_cmd=show run

0.8
===
+ Julien Louis added Dell.pm for Dell PowerConnect switches.

0.7
===
+ added Riverstone.pm for Alcatel Ethernet Routers.

0.6
===
+ added Parks.pm.
+ log program version. (cmd line option: -v)

0.5
===
+ added show_cmd option for older routers in CiscoIOS.pm. (Task 6204)
+ added FortiGate.pm.  (Doug Schaapveld / Task 6205)
+ modified CiscoIOS.pm, FortiGate.pm to strip leading "show" commands
  and other info.  This results in clean configs that can be used
  as-is to restore a device.  (Doug Schaapveld)
+ added ProCurve.pm for HP Procurve switches.  (Doug Schaapveld)

0.4
===
+ added CiscoCAT.pm contributed by Marcos Sandez.

0.3
===
+ ported to Windows.

0.2
===
+ new filesystem layout.
+ FSF address updated.

0.1
===
+ notes about the device id.
+ copyright notice at beginning of source files.
+ added this CHANGES file.


fetchconfig.pl
#! /usr/bin/perl -w
#
# fetchconfig - Retrieving configuration for multiple devices
# Copyright (C) 2006 Everton da Silva Marques
#
# fetchconfig is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# fetchconfig is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with fetchconfig; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
# MA 02110-1301 USA.
#
# $Id: fetchconfig.pl,v 1.7 2007/07/20 17:05:39 evertonm Exp $

use strict;
use fetchconfig::Logger;
use fetchconfig::Constants;
use fetchconfig::model::Detector;

sub basename {
    my ($path) = @_;

    my @list;

    if ($^O eq 'MSWin32') {
	@list = split /\\/, $path;
    }
    else {
	@list = split /\//, $path;
    }

    pop @list;
}

my $me = basename($0);

my $log = fetchconfig::Logger->new({ prefix => $me });

$log->info('version ' . fetchconfig::Constants::version);

my @device_file_list;

foreach my $opt (@ARGV) {
    if ($opt eq '-v') {
	exit; # only show version
    }
    if ($opt =~ /^-devices=(.+)$/) {
	push @device_file_list, $1;
	next;
    }
    $log->error("unexpected argument: $opt");
    &usage;
    die;
}

if (@device_file_list < 1) {
    $log->error("at least one device list file is required");
    &usage;
    die;
}

fetchconfig::model::Detector->init($log);

foreach my $dev_file (@device_file_list) {
    &load_device_list($dev_file);
}

$log->info("done");

exit;

sub usage {
    warn "usage: $me [-v] -devices=file\n";
}

sub load_device_list {
    my ($filename) = @_;
    
    local *IN;
    
    if (!open(IN, "<$filename")) {
	$log->error("could not read device list: $filename: $!");
	return;
    }
    
    $log->debug("loading device list: $filename");

    my $line_num = 0;
    
    while (<IN>) {
	chomp;

	++$line_num;

	#$log->debug("[$line_num] $_");

	next if (/^\s*(#|$)/);

        fetchconfig::model::Detector->parse($filename, $line_num, $_);
    }
		 
    close IN;
}


fetchconfig/Logger.pm
# fetchconfig - Retrieving configuration for multiple devices
# Copyright (C) 2006 Everton da Silva Marques
#
# fetchconfig is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# fetchconfig is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with fetchconfig; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
# MA 02110-1301 USA.
#
# $Id$

package fetchconfig::Logger; # fetchconfig/Logger.pm

use strict;
use warnings;

sub new {
    my ($proto, $options) = @_;
    my $class = ref($proto) || $proto;

    # defaults
    my $self = {
	          prefix => '?'
		};

    # user options
    if (@_ > 1) {
	foreach (keys %$options) {
	    $self->{$_} = $options->{$_};
	}
    }

    bless $self, $class;
}

sub prefix {
    my ($self, $prefix) = @_;

    my $old = $self->{prefix};

    if (defined($prefix)) {
	$self->{prefix} = $prefix;
    }

    $old;
}

sub info {
    my ($self, $msg) = @_;

    warn $self->{prefix}, ": info: ", $msg, "\n";
}

sub debug {
    my ($self, $msg) = @_;

    warn $self->{prefix}, ": debug: ", $msg, "\n";
}

sub error {
    my ($self, $msg) = @_;

    warn $self->{prefix}, ": error: ", $msg, "\n";
}

1;


fetchconfig/Constants.pm
# fetchconfig - Retrieving configuration for multiple devices
# Copyright (C) 2007 Everton da Silva Marques
#
# fetchconfig is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# fetchconfig is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with fetchconfig; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
# MA 02110-1301 USA.
#
# $Id: Constants.pm,v 1.16 2011/06/20 20:48:41 evertonm Exp $

package fetchconfig::Constants; # fetchconfig/Constants.pm

use strict;
use warnings;

my $version = '0.22';

sub version {
    $version;
}

1;


fetchconfig/model/Dell.pm
# fetchconfig - Retrieving configuration for multiple devices
# Copyright (C) 2007 Julien Louis
# Copyright (C) 2006 Everton da Silva Marques
#
# fetchconfig is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# fetchconfig is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with fetchconfig; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
# MA 02110-1301 USA.
#
# $Id: Dell.pm,v 1.1 2007/07/31 12:12:31 evertonm Exp $

package fetchconfig::model::Dell; # fetchconfig/model/Dell.pm

use strict;
use warnings;
use Net::Telnet;
use fetchconfig::model::Abstract;

@fetchconfig::model::Dell::ISA = qw(fetchconfig::model::Abstract);

####################################
# Implement model::Abstract - Begin
#

sub label {
    'dell';
}

# "sub new" fully inherited from fetchconfig::model::Abstract

sub fetch {
    my ($self, $file, $line_num, $line, $dev_id, $dev_host, $dev_opt_tab) = @_;

    my $saved_prefix = $self->{log}->prefix; # save log prefix
    $self->{log}->prefix("$saved_prefix: dev=$dev_id host=$dev_host");

    my @conf = $self->do_fetch($file, $line_num, $line, $dev_id, $dev_host, $dev_opt_tab);

    # restore log prefix
    $self->{log}->prefix($saved_prefix);

    @conf;
}

#
# Implement model::Abstract - End
##################################

sub chat_login {
    my ($self, $t, $dev_id, $dev_host, $dev_opt_tab) = @_;
    my $ok;

    my ($prematch, $match) = $t->waitfor(Match => '/(User Name:|Password:) ?$/');
    if (!defined($prematch)) {
	$self->log_error("could not find login prompt");
	return undef;
    }

    $self->log_debug("found login prompt: [$match]");

    if ($match =~ /^User Name/) {
	my $dev_user = $self->dev_option($dev_opt_tab, "user");
	if (!defined($dev_user)) {
	    $self->log_error("login username needed but not provided");
	    return undef;
	}

	$ok = $t->print($dev_user);
	if (!$ok) {
	    $self->log_error("could not send login username");
	    return undef;
	}

	($prematch, $match) = $t->waitfor(Match => '/Password: ?$/');
	if (!defined($prematch)) {
	    $self->log_error("could not find password prompt");
	    return undef;
	}

	$self->log_debug("found password prompt: [$match]");
    }

    if ($match =~ /^Password/) {
	my $dev_pass = $self->dev_option($dev_opt_tab, "pass");
        if (!defined($dev_pass)) {
	    $self->log_error("login password needed but not provided");
	    return undef;
        }
	$ok = $t->print("$dev_pass");
	if (!$ok) {
	    $self->log_error("could not send login password");
	    return undef;
	}

        ($prematch, $match) = $t->waitfor(Match => '/(\S+)[>#] ?$/');
	if (!defined($prematch)) {
	    $self->log_error("could not find command prompt");
	    return undef;
	}

	$self->log_debug("found command prompt: [$match]");
    }

    if ($match !~ /^(\S+)\# ?$/) {
	$self->log_error("could not match enable command prompt");
	return undef;
    }

    my $prompt = $1;

    $self->{prompt} = $prompt; # save prompt

    $self->log_debug("logged in prompt=[$prompt]");

    $prompt;
}

sub expect_enable_prompt {
    my ($self, $t, $prompt) = @_;

    if (!defined($prompt)) {
	$self->log_error("internal failure: undefined command prompt");
	return undef;
    }

    my $enable_prompt_regexp = '/' . $prompt . '# ?$/';

    my ($prematch, $match) = $t->waitfor(Match => $enable_prompt_regexp);
    if (!defined($prematch)) {
	$self->log_error("could not match enable command prompt: $enable_prompt_regexp");
    }

    ($prematch, $match);
}

sub chat_fetch {
    my ($self, $t, $dev_id, $dev_host, $prompt, $fetch_timeout, $show_cmd, $conf_ref) = @_;
    my $ok;
    
    $ok = $t->print('term datadump');
    if (!$ok) {
	$self->log_error("could not send pager disabling command");
	return 1;
    }

    my ($prematch, $match) = $self->expect_enable_prompt($t, $prompt);
    return unless defined($prematch);

    my $full_show_cmd="show run";
    if(defined($show_cmd)) {
      if($show_cmd eq "wrterm") {
        $full_show_cmd="write term";
      }
    }

    $ok = $t->print($full_show_cmd);
    if (!$ok) {
	$self->log_error("could not send show run command: $full_show_cmd");
	return 1;
    }

	# Prevent "show run" command from appearing in config dump
	$t->getline();

	# Commment out garbage at top so config file can be restored
	# cleanly at a later date
	my($line,$top_info);
	while($line=$t->getline()) {
		# Failsafe: Just in case 'Current configuration'
		# doesn't appear, assume config begins with 'version'
		# or first valid comment.
		if($line=~/^version / || $line=~/^\!/) {
			$top_info.=$line;
			last;
		} else {
			$top_info .= $line;
		}
		# Normally, finding the "Current configuration" line
		# will be enough to exit this loop.
		last if $line=~/^Current configuration/;
	}

    my $save_timeout;
    if (defined($fetch_timeout)) {
        $save_timeout = $t->timeout;
        $t->timeout($fetch_timeout);
    }

    ($prematch, $match) = $self->expect_enable_prompt($t, $prompt);
    if (!defined($prematch)) {
	$self->log_error("could not find end of configuration");
	return 1;
    }

    if (defined($fetch_timeout)) {
        $t->timeout($save_timeout);
    }

    $self->log_debug("found end of configuration: [$match]");

    @$conf_ref = split /\n/, $top_info . $prematch;

    $self->log_debug("fetched: " . scalar @$conf_ref . " lines");

    return undef;
}

sub do_fetch {
    my ($self, $file, $line_num, $line, $dev_id, $dev_host, $dev_opt_tab) = @_;

    $self->log_debug("trying");

    my $dev_repository = $self->dev_option($dev_opt_tab, "repository");
    if (!defined($dev_repository)) {
	$self->log_error("undefined repository");
	return;
    }

    if (! -d $dev_repository) {
	$self->log_error("not a directory repository=$dev_repository at file=$file line=$line_num: $line");
	return;
    }

    if (! -w $dev_repository) {
	$self->log_error("unable to write to repository=$dev_repository at file=$file line=$line_num: $line");
	return;
    }

    my $dev_timeout = $self->dev_option($dev_opt_tab, "timeout");

    my $t = new Net::Telnet(Errmode => 'return', Timeout => $dev_timeout);

    my $ok = $t->open($dev_host);
    if (!$ok) {
	$self->log_error("could not connect: $!");
	return;
    }

    $self->log_debug("connected");

    my $prompt = $self->chat_login($t, $dev_id, $dev_host, $dev_opt_tab);

    return unless defined($prompt);

    my @config;

    my $fetch_timeout = $self->dev_option($dev_opt_tab, "fetch_timeout");

    my $show_cmd = $self->dev_option($dev_opt_tab, "show_cmd");

    return if $self->chat_fetch($t, $dev_id, $dev_host, $prompt, $fetch_timeout, $show_cmd, \@config);

    $ok = $t->close;
    if (!$ok) {
	$self->log_error("disconnecting: $!");
    }

    $self->log_debug("disconnected");

    $self->dump_config($dev_id, $dev_opt_tab, \@config);
}

1;



fetchconfig/model/CiscoIOS.pm
# fetchconfig - Retrieving configuration for multiple devices
# Copyright (C) 2006 Everton da Silva Marques
#
# fetchconfig is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# fetchconfig is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with fetchconfig; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
# MA 02110-1301 USA.
#
# $Id: CiscoIOS.pm,v 1.7 2007/08/14 21:19:08 evertonm Exp $

package fetchconfig::model::CiscoIOS; # fetchconfig/model/CiscoIOS.pm

use strict;
use warnings;
use Net::Telnet;
use fetchconfig::model::Abstract;

@fetchconfig::model::CiscoIOS::ISA = qw(fetchconfig::model::Abstract);

####################################
# Implement model::Abstract - Begin
#

sub label {
    'cisco-ios';
}

# "sub new" fully inherited from fetchconfig::model::Abstract

sub fetch {
    my ($self, $file, $line_num, $line, $dev_id, $dev_host, $dev_opt_tab) = @_;

    my $saved_prefix = $self->{log}->prefix; # save log prefix
    $self->{log}->prefix("$saved_prefix: dev=$dev_id host=$dev_host");

    my @conf = $self->do_fetch($file, $line_num, $line, $dev_id, $dev_host, $dev_opt_tab);

    # restore log prefix
    $self->{log}->prefix($saved_prefix);

    @conf;
}

#
# Implement model::Abstract - End
##################################

sub chat_login {
    my ($self, $t, $dev_id, $dev_host, $dev_opt_tab) = @_;
    my $ok;

    my $login_prompt = '/(Username:|Password:) $/';

    # chat_banner is used to allow temporary modification
    # of timeout throught the 'banner_timeout' option

    my ($prematch, $match) = $self->chat_banner($t, $dev_opt_tab, $login_prompt);
    if (!defined($prematch)) {
	$self->log_error("could not find login prompt: $login_prompt");
	return undef;
    }

    $self->log_debug("found login prompt: [$match]");

    if ($match =~ /^Username/) {
	my $dev_user = $self->dev_option($dev_opt_tab, "user");
	if (!defined($dev_user)) {
	    $self->log_error("login username needed but not provided");
	    return undef;
	}

	$ok = $t->print($dev_user);
	if (!$ok) {
	    $self->log_error("could not send login username");
	    return undef;
	}

	($prematch, $match) = $t->waitfor(Match => '/Password: $/');
	if (!defined($prematch)) {
	    $self->log_error("could not find password prompt");
	    return undef;
	}

	$self->log_debug("found password prompt: [$match]");
    }

    if ($match =~ /^Password/) {
	my $dev_pass = $self->dev_option($dev_opt_tab, "pass");
        if (!defined($dev_pass)) {
	    $self->log_error("login password needed but not provided");
	    return undef;
        }

	$ok = $t->print($dev_pass);
	if (!$ok) {
	    $self->log_error("could not send login password");
	    return undef;
	}

        ($prematch, $match) = $t->waitfor(Match => '/(\S+)[>#]$/');
	if (!defined($prematch)) {
	    $self->log_error("could not find command prompt");
	    return undef;
	}

	$self->log_debug("found command prompt: [$match]");
    }

    if ($match =~ /^\S+>$/) {
        $ok = $t->print('enable');
	if (!$ok) {
	    $self->log_error("could not send enable command");
	    return undef;
	}
	
        ($prematch, $match) = $t->waitfor(Match => '/(Password: |\S+#)$/');
	if (!defined($prematch)) {
	    $self->log_error("could not find enable password prompt");
	    return undef;
	}

        if ($match =~ /^Password/) {
	    my $dev_enable = $self->dev_option($dev_opt_tab, "enable");
	    if (!defined($dev_enable)) {
		$self->log_error("enable password needed but not provided");
		return undef;
	    }

	    $ok = $t->print($dev_enable);
	    if (!$ok) {
		$self->log_error("could not send enable password");
		return undef;
	    }

	    ($prematch, $match) = $t->waitfor(Match => '/\S+#$/');
	    if (!defined($prematch)) {
		$self->log_error("could not find enable command prompt");
		return undef;
	    }
        }

	$self->log_debug("found enable prompt: [$match]");
    }

    if ($match !~ /^(\S+)\#$/) {
	$self->log_error("could not match enable command prompt");
	return undef;
    }

    my $prompt = $1;

    $self->{prompt} = $prompt; # save prompt

    $self->log_debug("logged in prompt=[$prompt]");

    $prompt;
}

sub expect_enable_prompt {
    my ($self, $t, $prompt) = @_;

    if (!defined($prompt)) {
	$self->log_error("internal failure: undefined command prompt");
	return undef;
    }

    my $enable_prompt_regexp = '/' . $prompt . '#$/';

    my ($prematch, $match) = $t->waitfor(Match => $enable_prompt_regexp);
    if (!defined($prematch)) {
	$self->log_error("could not match enable command prompt: $enable_prompt_regexp");
    }

    ($prematch, $match);
}

sub chat_fetch {
    my ($self, $t, $dev_id, $dev_host, $prompt, $fetch_timeout, $show_cmd, $conf_ref) = @_;
    my $ok;
    
    $ok = $t->print('term len 0');
    if (!$ok) {
	$self->log_error("could not send pager disabling command");
	return 1;
    }

    my ($prematch, $match) = $self->expect_enable_prompt($t, $prompt);
    return unless defined($prematch);

    # Backward compatibility support for option "show_cmd=wrterm"
    my $custom_cmd;
    if (defined($show_cmd)) {
	$custom_cmd = ($show_cmd eq 'wrterm') ? 'write term' : $show_cmd;
    }

    if ($self->chat_show_conf($t, 'show run', $custom_cmd)) {
	return 1;
    }

	# Prevent "show run" command from appearing in config dump
	$t->getline();

	# Commment out garbage at top so config file can be restored
	# cleanly at a later date
	my($line,$top_info);
	while($line=$t->getline()) {
		# Failsafe: Just in case 'Current configuration'
		# doesn't appear, assume config begins with 'version'
		# or first valid comment.
		if($line=~/^version / || $line=~/^\!/) {
			$top_info.=$line;
			last;
		} else {
			$top_info.='!!' . $line;
		}
		# Normally, finding the "Current configuration" line
		# will be enough to exit this loop.
		last if $line=~/^Current configuration/;
	}

    my $save_timeout;
    if (defined($fetch_timeout)) {
        $save_timeout = $t->timeout;
        $t->timeout($fetch_timeout);
    }

    ($prematch, $match) = $self->expect_enable_prompt($t, $prompt);
    if (!defined($prematch)) {
	$self->log_error("could not find end of configuration");
	return 1;
    }

    if (defined($fetch_timeout)) {
        $t->timeout($save_timeout);
    }

    $self->log_debug("found end of configuration: [$match]");

    @$conf_ref = split /\n/, $top_info . $prematch;

    $self->log_debug("fetched: " . scalar @$conf_ref . " lines");

    return undef;
}

sub do_fetch {
    my ($self, $file, $line_num, $line, $dev_id, $dev_host, $dev_opt_tab) = @_;

    $self->log_debug("trying");

    my $dev_repository = $self->dev_option($dev_opt_tab, "repository");
    if (!defined($dev_repository)) {
	$self->log_error("undefined repository");
	return;
    }

    if (! -d $dev_repository) {
	$self->log_error("not a directory repository=$dev_repository at file=$file line=$line_num: $line");
	return;
    }

    if (! -w $dev_repository) {
	$self->log_error("unable to write to repository=$dev_repository at file=$file line=$line_num: $line");
	return;
    }

    my $dev_timeout = $self->dev_option($dev_opt_tab, "timeout");

    my $t = new Net::Telnet(Errmode => 'return', Timeout => $dev_timeout);

    my $ok = $t->open($dev_host);
    if (!$ok) {
	$self->log_error("could not connect: $!");
	return;
    }

    $self->log_debug("connected");

    my $prompt = $self->chat_login($t, $dev_id, $dev_host, $dev_opt_tab);

    return unless defined($prompt);

    my @config;

    my $fetch_timeout = $self->dev_option($dev_opt_tab, "fetch_timeout");

    my $show_cmd = $self->dev_option($dev_opt_tab, "show_cmd");

    return if $self->chat_fetch($t, $dev_id, $dev_host, $prompt, $fetch_timeout, $show_cmd, \@config);

    $ok = $t->close;
    if (!$ok) {
	$self->log_error("disconnecting: $!");
    }

    $self->log_debug("disconnected");

    $self->dump_config($dev_id, $dev_opt_tab, \@config);
}

1;
